# Day 7: No Space Left On Device

[https://adventofcode.com/2022/day/7](https://adventofcode.com/2022/day/7)

## Description

### Part One

You can hear birds chirping and raindrops hitting leaves as the expedition proceeds. Occasionally, you can even hear much louder sounds in the distance; how big do the animals get out here, anyway?

The device the Elves gave you has problems with more than just its communication system. You try to run a system update:

    $ system-update --please --pretty-please-with-sugar-on-top
    Error: No space left on device
    

Perhaps you can delete some files to make space for the update?

You browse around the filesystem to assess the situation and save the resulting terminal output (your puzzle input). For example:

    $ cd /
    $ ls
    dir a
    14848514 b.txt
    8504156 c.dat
    dir d
    $ cd a
    $ ls
    dir e
    29116 f
    2557 g
    62596 h.lst
    $ cd e
    $ ls
    584 i
    $ cd ..
    $ cd ..
    $ cd d
    $ ls
    4060174 j
    8033020 d.log
    5626152 d.ext
    7214296 k
    

The filesystem consists of a tree of files (plain data) and directories (which can contain other directories or files). The outermost directory is called `/`. You can navigate around the filesystem, moving into or out of directories and listing the contents of the directory you're currently in.

Within the terminal output, lines that begin with `$` are _commands you executed_, very much like some modern computers:

*   `cd` means _change directory_. This changes which directory is the current directory, but the specific result depends on the argument:
    *   `cd x` moves _in_ one level: it looks in the current directory for the directory named `x` and makes it the current directory.
    *   `cd ..` moves _out_ one level: it finds the directory that contains the current directory, then makes that directory the current directory.
    *   `cd /` switches the current directory to the outermost directory, `/`.
*   `ls` means _list_. It prints out all of the files and directories immediately contained by the current directory:
    *   `123 abc` means that the current directory contains a file named `abc` with size `123`.
    *   `dir xyz` means that the current directory contains a directory named `xyz`.

Given the commands and output in the example above, you can determine that the filesystem looks visually like this:

    - / (dir)
      - a (dir)
        - e (dir)
          - i (file, size=584)
        - f (file, size=29116)
        - g (file, size=2557)
        - h.lst (file, size=62596)
      - b.txt (file, size=14848514)
      - c.dat (file, size=8504156)
      - d (dir)
        - j (file, size=4060174)
        - d.log (file, size=8033020)
        - d.ext (file, size=5626152)
        - k (file, size=7214296)
    

Here, there are four directories: `/` (the outermost directory), `a` and `d` (which are in `/`), and `e` (which is in `a`). These directories also contain files of various sizes.

Since the disk is full, your first step should probably be to find directories that are good candidates for deletion. To do this, you need to determine the _total size_ of each directory. The total size of a directory is the sum of the sizes of the files it contains, directly or indirectly. (Directories themselves do not count as having any intrinsic size.)

The total sizes of the directories above can be found as follows:

*   The total size of directory `e` is _584_ because it contains a single file `i` of size 584 and no other directories.
*   The directory `a` has total size _94853_ because it contains files `f` (size 29116), `g` (size 2557), and `h.lst` (size 62596), plus file `i` indirectly (`a` contains `e` which contains `i`).
*   Directory `d` has total size _24933642_.
*   As the outermost directory, `/` contains every file. Its total size is _48381165_, the sum of the size of every file.

To begin, find all of the directories with a total size of _at most 100000_, then calculate the sum of their total sizes. In the example above, these directories are `a` and `e`; the sum of their total sizes is _`95437`_ (94853 + 584). (As in this example, this process can count files more than once!)

Find all of the directories with a total size of at most 100000. _What is the sum of the total sizes of those directories?_


In [17]:
from pathlib import Path
from typing import Generator

TEST = """$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k""".splitlines()

EXPECTED_1 = 95437
EXPECTED_2 = 24933642

DATADIR = Path("input")
DATA = (DATADIR / "data7.txt").read_text().splitlines()

from collections import defaultdict


def parse(data: list[str]) -> dict:
    path = Path("/")
    files = {}
    cwd = files
    for row in data:
        if row == "$ cd /":
            path = Path("/")
        if row.startswith("$ cd"):
            path = (path / row.split()[-1]).resolve(strict=False)
            cwd = files
            for dir in str(path).split("/"):
                if dir:
                    cwd = cwd.setdefault(dir, {})
        elif row.startswith("$ ls"):
            continue
        elif row.startswith("dir "):
            cwd[row.split()[-1]] = {}
        else:
            size, name = row.split()
            cwd[name] = int(size)
    return files


<div class="alert alert-info">Perhaps not needed but I parsed the 'commands' into a nested dictionary of filenames and sizes or subdirectories. I guess I could have just added things up in the parsing but at this point I didn't know what I might need for part 2.</div>

In [18]:
from pprint import pprint
pprint(parse(TEST))

{'a': {'e': {'i': 584}, 'f': 29116, 'g': 2557, 'h.lst': 62596},
 'b.txt': 14848514,
 'c.dat': 8504156,
 'd': {'d.ext': 5626152, 'd.log': 8033020, 'j': 4060174, 'k': 7214296}}


In [19]:
def sizes(files: dict) -> list[int]:
    total = 0
    all_sub = []
    for name, v in files.items():
        if isinstance(v, dict):
            sub = sizes(v)
            all_sub += sub
            if sub:
                total += sub[-1]
        else:
            total += v
    return all_sub + [total]



<div class="alert alert-info">Second step is to walk the dict and add up the sizes of files and directories it contains. Last element of the list is the total size of the directory, earlier results are the sizes of its subdirectories.</div>

In [20]:
pprint(sizes(parse(TEST)))

[584, 94853, 24933642, 48381165]


In [21]:

def score_1(data: list[str]) -> int:
    files = parse(data)
    return sum(f for f in sizes(files) if f < 100000)


test_score = score_1(TEST)
print("test", test_score)
assert test_score == EXPECTED_1
print("part 1", score_1(DATA))


test 95437
part 1 1749646


<div class="alert alert-info">Part 1 is just add up the sizes of all the smaller directories</div>


### Part Two

Now, you're ready to choose a directory to delete.

The total disk space available to the filesystem is _`70000000`_. To run the update, you need unused space of at least _`30000000`_. You need to find a directory you can delete that will _free up enough space_ to run the update.

In the example above, the total size of the outermost directory (and thus the total amount of used space) is `48381165`; this means that the size of the _unused_ space must currently be `21618835`, which isn't quite the `30000000` required by the update. Therefore, the update still requires a directory with total size of at least `8381165` to be deleted before it can run.

To achieve this, you have the following options:

*   Delete directory `e`, which would increase unused space by `584`.
*   Delete directory `a`, which would increase unused space by `94853`.
*   Delete directory `d`, which would increase unused space by `24933642`.
*   Delete directory `/`, which would increase unused space by `48381165`.

Directories `e` and `a` are both too small; deleting them would not free up enough space. However, directories `d` and `/` are both big enough! Between these, choose the _smallest_: `d`, increasing unused space by _`24933642`_.

Find the smallest directory that, if deleted, would free up enough space on the filesystem to run the update. _What is the total size of that directory?_


In [22]:
def score_2(data: list[str]) -> int:
    disk_size = 70000000
    needed = 30000000
    files = parse(data)

    sz = sizes(files)
    used = sz[-1]
    unused = disk_size - used
    needed -= unused
    return min(size for size in sz if size >= needed)


if EXPECTED_2 is not None:
    test_score = score_2(TEST)
    print("test", test_score)
    assert test_score == EXPECTED_2
    print("part 2", score_2(DATA))


test 24933642
part 2 1498966


<div class="alert alert-info">It turns out all I needed for part 2 was the same list of sizes as in part 1 and return the smallest directory size greater than the amount needed.</div>